classes/ratings_and_allocations_table.php

```diff
@@ -553,38 +572,92 @@ private function filter_userids($userids) {
         if (!$userids) {
             return $userids;
         }
-        if (!$this->hidenorating && !$this->showallocnecessary) {
+        if (!$this->hidenorating && !$this->showallocnecessary && $this->groupselect == 0) {
             return $userids;
         }
         $sql = "SELECT distinct u.id FROM {user} u ";
+
         if ($this->hidenorating) {
             $sql .= "JOIN {ratingallocate_ratings} r ON u.id=r.userid " .
-                    "JOIN {ratingallocate_choices} c ON r.choiceid = c.id " .
-                    "AND c.ratingallocateid = :ratingallocateid " .
-                    "AND c.active=1 ";
+                "JOIN {ratingallocate_choices} c ON r.choiceid = c.id " .
+                "AND c.ratingallocateid = :ratingallocateid " .
+                "AND c.active=1 ";
         }
         if ($this->showallocnecessary) {
             $sql .= "LEFT JOIN ({ratingallocate_allocations} a " .
-                    "JOIN {ratingallocate_choices} c2 ON c2.id = a.choiceid AND c2.active=1 " .
-                    "AND a.ratingallocateid = :ratingallocateid2 )" .
-                    "ON u.id=a.userid " .
-                    "WHERE a.id is null AND u.id in (" . implode(",", $userids) . ") ";
+                "JOIN {ratingallocate_choices} c2 ON c2.id = a.choiceid AND c2.active=1 " .
+                "AND a.ratingallocateid = :ratingallocateid2 )" .
+                "ON u.id=a.userid " .
+                "WHERE a.id is null AND u.id in (" . implode(",", $userids) . ") ";
         } else {
             $sql .= "WHERE u.id in (" . implode(",", $userids) . ") ";
         }
+        if ($this->groupselect == -1) {
+            $sql .= "AND u.id not in ( SELECT distinct gm.userid FROM {groups_members} gm WHERE gm.groupid in (null";
+            if (!empty($gmgroupid = implode(",",
+                array_map(
+                    function($o) {
+                        return $o->id;
+                    },
+                    $this->groupsofallchoices)))) {
+                $sql .= "," . $gmgroupid . ") ) ";
+            }
+        } else if ($this->groupselect != 0) {
+            $sql .= "AND u.id in ( SELECT gm.userid FROM {groups_members} gm WHERE gm.groupid= :groupselect ) ";
+        }
         return array_map(
                 function($u) {
                     return $u->id;
                 },
                 $DB->get_records_sql($sql,
                         array(
                                 'ratingallocateid' => $this->ratingallocate->ratingallocate->id,
-                                'ratingallocateid2' => $this->ratingallocate->ratingallocate->id
+                                'ratingallocateid2' => $this->ratingallocate->ratingallocate->id,
```

Comment:
You don't need a second duplicate variable here, just reuse the existing ratingallocateid.